### PR TITLE
Project Branching Documentation

### PR DESCRIPTION
**Pull request by mlavin**
This is adapted from previous documentation written by @robline. Feedback is welcome. If you find a typo/grammatical error then please just fix it rather than commenting.

### CODE REVIEW

**Comment by mlavin** [comment]
Updated based on feedback and included a new section on production releases. Please take another look.

File: source/branching.rst [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+-----------------
+
+Recall that ``origin/develop`` and ``origin/master`` are the only long term branches. Every other
+branch should be as short-lived as possible while still accomplishing its purpose. Dead branches
+(exploratory, or failed attempts to address an issue), or feature branches that drag on a long time,
+cloud the picture for other developers. Since this feature branch is no longer needed, it can be deleted
+via the "delete this branch" button on GitHub, or via the command line:
+
+    git branch -d 42-my-new-feature
+
```
Comment (mlavin) [inline source/branching.rst]: I've noted that the use of `--no-ff` in the merge can be used to preserve the original branch information in the commit.

File: source/branching.rst [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+It is also incumbent upon the pull request reviewer to read the code diff and make comments on the code.
+Github presents a blue and white plus icon for adding line comments.
+
+When the reviewer agrees that the pull request is adequate, they should add a comment to the pull request.
+For fun, we typically add random emoticons such as ``:frog:``.
+
+By convention, the developer of the branch then merges it back into ``develop``, but other developers can do this step::
+
+    $ git checkout develop
+    $ git merge --no-ff 42-my-new-feature
```
Comment (ironfroggy) [inline source/branching.rst]: I think it is worth mentioning the `--no-ff` flag and the reason its recommended.

**Comment by ironfroggy** [comment]
:gem:

**Comment by mlavin** [comment]
@rebecca-caktus @dpoirier @kmtracey do you feel like your feedback has been addressed/incorporated?

File: source/branching.rst [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+Branch Lifespans
+-----------------
+
+Recall that ``origin/develop`` and ``origin/master`` are the only long term branches. Every other
+branch should be as short-lived as possible while still accomplishing its purpose. Dead branches
+(exploratory, or failed attempts to address an issue), or feature branches that drag on a long time,
+cloud the picture for other developers. Since this feature branch is no longer needed, it can be deleted
+via the "delete this branch" button on GitHub, or via the command line::
+
+    $ git branch -d 42-my-new-feature
```
Comment (dpoirier) [inline source/branching.rst]: This command only deletes the branch from the local copy of the repo, not from GitHub.  There's some arcane command line way to delete branches from a remote repo, but I just use the branches view on GitHub to remove branches that have outlived their usefulness.

File: source/branching.rst [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+This merge should be a no-op. If there are changes which were not previously merged to develop you need
+to ensure that develop is still running as expected before preceeding. While on the ``develop`` branch
+you should 
+
+- Update change log with release notes
+- Update internal version numbers if tracked in the code
+- Run test suite
+
+Changes between each release should be noted in a ``CHANGELOG`` or ``CHANGES`` file stored at the root of the repository.
+The format of this file is flexible. You may look at https://github.com/caktus/margarita/blob/develop/CHANGES.rst
```
Comment (dpoirier) [inline source/branching.rst]: or if the project has Sphinx docs, we often will document releases in a file like `docs/changes.rst` or `docs/release_notes.rst`.

File: source/branching.rst (hunk 3 above)
Comment (mlavin) [inline source/branching.rst]: Yes I believe most use the Github UI to delete the remote branches. Is it worth adding a note regarding that or including the command to delete the remote branch `git push origin :42-my-new-feature`?

Comment (dpoirier) [inline source/branching.rst]: I think it's at least worth noting that this is the command to clean up your local repo but it won't clean up GitHub.

**Comment by mlavin** [comment]
Updated again.

**Comment by dpoirier** [comment]
:umbrella:

**Pull request merged**